### ISSUE TITLE
the git-hooks can not work on windows(mingw)

### ISSUE THREAD

**Issue by snowyu** [tarmolov/git-hooks-js#10]
``` bash
[GIT-HOOKS ERROR] Cannot find module '....
node_modulesgit-hookslib/git-hooks'
```

the path delimiter of the generated hook files is error:

.git/hooks/post-commit:

``` bash
#!/usr/bin/env node

try {
    require('..\..\node_modules\git-hooks\lib/git-hooks').run(__filename, process.argv[2], function (code) {
        process.exit(code);
    });
} catch (e) {
    console.error('[GIT-HOOKS ERROR] ' + e.message);
}
```

**Comment by tarmolov** [comment]
To be honest, I didn't test `git-hooks` on Windows but if you have any time to make pr, I'll really appreciate it ;)

See also https://github.com/tarmolov/git-hooks-js/pull/7#issuecomment-135899560

**Comment by aliosv** [comment]
Current problems:
1) wrong path to git-hooks in template
2) checking file on the execution not works in windows
3) windows not allow shebangs - so here problems with spawn
here fast fix https://github.com/tarmolov/git-hooks-js/pull/11

**Comment by tarmolov** [comment]
Could you tell me in details about each of problem please?

**Comment by aliosv** [comment]
1) https://github.com/tarmolov/git-hooks-js/blob/master/lib/hook-template.js#L4 -> wrong path '..\..\lib/git-hooks'
2) https://github.com/tarmolov/git-hooks-js/blob/master/lib/git-hooks.js#L148 always false 
3) default windows executable files is .bat or .cmd. You can associate file extensions with executable programs(script.bash -> execute by bash), but we use shebangs instead of extensions, which windows not support.

**Comment by tarmolov** [comment]
1) `path.join` should fix this problem.
2) It seems that for win32 should be always returned `true` https://github.com/kevva/executable/blob/master/index.js#L7-L9
3) spawn for windows is trickier than you think https://github.com/IndigoUnited/node-cross-spawn-async but handling shebang is probably enough for the majority cases. And again it's better to store all windows related stuff in a separate file. Just for the record, `shebang-regex` is just [one line module](https://github.com/sindresorhus/shebang-regex/blob/master/index.js#L2) and `shebang-command` is [7 lines long](https://github.com/kevva/shebang-command/blob/master/index.js#L5-L14).

**Comment by aliosv** [comment]
1) path.join returns '..\..\lib\git-hooks', in hook-template you need to have '..\\..\lib\\git-hooks', so you also need to escape backslashes.
2) if you agree with shebangs, why ont to leave current check(if file not contain, then it is not executable for windows)?
3) i will get rid of dependencies, about separate file i wrote earlier: explain in details, i don't see what to move(spawnHook function?)

**Comment by cinderblock** [comment]
Why not just use POSIX-compliant paths when on win32? Get rid of all those pesky backslashes asap.

I'm creating a PR that does this. It also fixes an endless loop that can occur on win32 systems.

Why not use cross-spawn-async to fix all of our spawn issues?

**Comment by cinderblock** [comment]
Actually, why is the path to git-hooks even used in the hooks? Pushing another commit shortly.

**Comment by tarmolov** [comment]
`cross-spawn-async` isn't used because of npm@2 bug https://github.com/npm/npm/issues/8850

**Comment by Loque-** [comment]
The static path being added here also prevents hooks from running successfully when using a git gui (as the gui does not run from the same place). Would it be possible for it to work out the path of the project with something like `require('path').resolve(__dirname);` ?

**Comment by tarmolov** [comment]
Sorry. I didn't get you.
Do you mean that https://github.com/tarmolov/git-hooks-js/blob/master/lib/hook-template.js#L11 doesn't work?

**Comment by Loque-** [comment]
I had not seen that update, looks exactly like the upgrade required (I have no idea what is available to git hooks). Thanks so much!

**Comment by jvanoostveen** [comment]
In Powershell I see this error:
```
[GIT-HOOKS ERROR] Cannot execute hook: C:\Users\jvo\Projects\ProjectX\.githooks\pre-commit\gulp.cmd. Please check file permissions.
```

**Comment by jvanoostveen** [comment]
We are using our own fork right now, but it would be nice to have it working from this repository.
The hack we did: https://github.com/AFASSoftware/git-hooks-js/commit/2deec719815d348410c04ea5a39dc322b5c8772a

**Comment by tarmolov** [comment]
Feel free to make a pr ;)

**Comment by codebling** [comment]
The spawn issue is actually very easy to solve. Since we know Git is installed on the machine, and Git has a copy of `sh`, just spawn `sh` and let it determine how to run the executables. This is what Git does for its own githooks and so is a good standardised way of handling the issue. 

The code is already available in PR #29, I'm just fixing all of the tests right now so that they work on Windows.

**Comment by codebling** [comment]
There are a few PRs that offer Windows support (see #11, #29, #38, #52 and your own #51) but it's unlikely that these will be merged, see tarmolov's comment [here](https://github.com/tarmolov/git-hooks-js/pull/38#issuecomment-271422549). Check out [Husky](https://github.com/typicode/husky), a very active project which allows you to set up git hooks in your package.json file. It supports OSX, linux and Windows and has support for a few other hooks which git-hooks-js does not.
